File: templates/faastroute-python/rpc.py
from concurrent import futures
import os
import redis
import datetime
import grpc
import app_pb2 as pb
import app_pb2_grpc as pb_grpc

def RPC(dest, payloads, id):
    if "LOGGING_NAME" in os.environ:
        redisClient = redis.Redis(host=os.environ['LOGGING_URL'], password=os.environ['LOGGING_PASSWORD'])
    with grpc.insecure_channel(dest) as channel:
        stub = pb_grpc.gRPCFunctionStub(channel)
        tl = []
        if "LOGGING_NAME" in os.environ:
            redisClient.append(os.environ["LOGGING_NAME"], "INVOKE_START," + id + "," + dest + "," + str(len(payloads)) + "," + str(datetime.datetime.now()) + "\n")
        with futures.ThreadPoolExecutor(max_workers=len(payloads)) as executor:
            for payload in payloads:
                tl.append(executor.submit(stub.gRPCFunctionHandler, pb.RequestBody(body=bytes(payload, "utf-8"), workflow_id=id)))
        if "LOGGING_NAME" in os.environ:
            redisClient.append(os.environ["LOGGING_NAME"], "INVOKE_END," + id + "," + dest + "," + str(len(payloads)) + "," + str(datetime.datetime.now()) + "\n")
        results = [t.result().reply for t in tl]
        return results

File: templates/faastroute-go/rpc.go
package function

import (
    "os"
    "context"
    "time"
    "github.com/redis/go-redis/v9"
    "strconv"
    "google.golang.org/grpc"
    "google.golang.org/grpc/credentials/insecure"

    pb "function/app_pb"
)

func invoke(stub pb.GRPCFunctionClient, ctx context.Context, in *pb.RequestBody) (*pb.ResponseBody) {
    res, _ := stub.GRPCFunctionHandler(ctx, in)
    return res
}

func RPC(dest string, payloads []string, id string) ([]string) {
    logging_name, logging := os.LookupEnv("LOGGING_NAME")
    redisClient := redis.NewClient(&redis.Options{})
    c := context.Background()
    if logging {
        logging_url := os.Getenv("LOGGING_URL")
        logging_password := os.Getenv("LOGGING_PASSWORD")
        redisClient = redis.NewClient(&redis.Options{
            Addr: logging_url,
            Password: logging_password,
            DB: 0,
        })
    }
    channel, _ := grpc.Dial(dest, grpc.WithTransportCredentials(insecure.NewCredentials()))
    defer channel.Close()
    stub := pb.NewGRPCFunctionClient(channel)
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    defer cancel()
    tl := make(chan *pb.ResponseBody)
    if logging {
        redisClient.Append(c, logging_name, "INVOKE_START," + id + "," + dest + "," + strconv.Itoa(len(payloads)) + "," + time.Now().String() + "\n")
    }
    for i := 0; i < len(payloads); i++ {
        go func(n int) {
            tl <- invoke(stub, ctx, &pb.RequestBody{Body: []byte(payloads[n]), WorkflowId: id})
        }(i)
    }
    results := make([]string, 0)
    for i := 0; i < len(payloads); i++ {
        res := <-tl
        results = append(results, res.GetReply())
    }
    if logging {
        redisClient.Append(c, logging_name, "INVOKE_END," + id + "," + dest + "," + strconv.Itoa(len(payloads)) + "," + time.Now().String() + "\n")
    }
    return results
}
